File: katas/4 kyu/sum-by-factors/solution.py
# sum-by-factors
import codewars_test as test
# -------------------------------- MY SOLUTION ------------------------------- #
def sum_for_list(lst):
    PrimeLst = []
    FinalList = []
    for n in lst:
        PrimeLst = list(set(PrimeLst).union(set(prime_factors(n))))
    PrimeLst.sort()
    for n in PrimeLst:
        sum = 0
        listN = [n]
        for m in lst:
            if m%n ==0:
                sum = sum+m
        listN.append(sum)
        FinalList.append(listN)
    return FinalList

def prime_factors(n):
    i = 2
    n = abs(n)
    factors = []
    while i * i <= n:         
        if n % i:             
            i += 1        
        else:             
            n //= i             
            factors.append(i)     
    if n > 1:
        factors.append(n)
    return factors

# ------------------------------ CLEVER SOLUTION ----------------------------- #
# ----------------------------------- TEST ----------------------------------- #
a = [12, 15]
test.assert_equals(sum_for_list(a), [[2, 12], [3, 27], [5, 15]])
a = [15, 21, 24, 30, 45]
test.assert_equals(sum_for_list(a), [[2, 54], [3, 135], [5, 90], [7, 21]])

File: katas/4 kyu/sum-by-factors/solution.php
<!-- sum-by-factors -->
<?php
use PHPUnit\Framework\TestCase;
/* ------------------------------- MY SOLUTION ------------------------------ */
function sumOfDivided($lst) {
  if (!is_array($lst) || empty($lst)) return array();

  $max = max(max($lst), abs(min($lst)));
  
  //get primes
  $sqrtmax = (int)sqrt($max);
  $primes[2] = true;
  for ($i = 3; $i <= $max; $i+=2) {
    $primes[$i] = true;
  }
  
  for($i = 3; $i <= $sqrtmax; $i+=2) {
    $z = 2;
    while($z*$i <=$max) {
      unset($primes[$i*$z]);
      $z++;
    }
  }


  foreach($primes as $prime => $true) {
    $isdivided = 0;
    $result = 0;
    foreach($lst as $num) {
      if ($num%$prime == 0) {
        $isdivided = 1;
        $result+=$num;
      }
    }
    if ($isdivided) {
      $return[]=[$prime, $result];
    }
  }

return $return;

}
/* ----------------------------- CLEVER SOLUTION ---------------------------- */
/* ---------------------------------- TEST ---------------------------------- */
class SumOfDividedTest extends TestCase
{
    private function revTest($actual, $expected) {
        $this->assertSame($expected, $actual);
    }
    public function testBasics() {        
        $this->revTest(sumOfDivided([12, 15]), [ [2, 12], [3, 27], [5, 15] ]);
        $this->revTest(sumOfDivided([15,21,24,30,45]), [ [2, 54], [3, 135], [5, 90], [7, 21] ]);
        $this->revTest(sumOfDivided([15,21,24,30,-45]), [ [2, 54], [3, 45], [5, 0], [7, 21] ]);
    }
}
